src/pycsou/opt/solver/lsqr.py

```diff
@@ -0,0 +1,311 @@
+# lsqr.py
+# ========
+# Author : Kaan Okumus [okukaan@gmail.com]
+# #############################################################################
+
+import typing as typ
+
+import numpy as np
+
+import pycsou.abc.operator as pyco
+import pycsou.abc.solver as pycs
+import pycsou.opt.stop as pycos
+import pycsou.runtime as pycrt
+import pycsou.util as pycu
+import pycsou.util.ptype as pyct
+
+
+class LSQR(pycs.Solver):
+    r"""
+    LSQR Method.
+
+    The LSQR method solves the system of linear equations :math:`\mathbf{A}\mathbf{x}=\mathbf{b}` iteratively.
+    :math:`\mathbf{A}` is a rectangular matrix of dimension m-by-n, where all cases are allowed: m=n, m>n or m<n.
+    :math:`\mathbf{b}` is a vector of length m. The matrix :math:`\mathbf{A}` may be square or rectangular and may have
+    any rank.
+
+    * For unsymmetric equations, it solves :math:`\mathbf{A}\mathbf{x}=\mathbf{b}`.
+
+    * For linear least squares, it solves :math:`||\mathbf{b}-\mathbf{A}\mathbf{x}||^2`.
+
+    * For damped least squares, it solves :math:`||\mathbf{b}-\mathbf{A}\mathbf{x}||^2 + d^2 ||\mathbf{x}-\mathbf{x_0}||^2`.
+
+
+    ``LSQR()`` **Parameterization**
+
+    A: :py:class:`pycsou.abc.operator.PosDefOp`
+        Any positive definite linear operator is accepted.
+    damp: float
+        Damping coefficient. Default is 0.
+    atol, btol: float
+        Stopping tolerances. LSQR continues iterations until a certain backward error estimate is smaller than some
+        quantity depending on atol and btol. Default is 1e-6 for both.
+    conlim: float
+        LSQR terminates if an estimate :math:`\text{cond}(\mathbf{A})` exceeds conlim. Default is 1e8.
+    iter_lim: int, optional
+        LSQR terminates if the number of iterations reaches iter_lim. Default is None.
+
+    ``LSQR.fit()`` **Parameterization**
+
+    b: NDArray
+         (..., N) 'b' terms in the LSQR cost function. All problems are solved in parallel.
+    x0: NDArray
+        (..., N) initial point(s). Defaults to 0 if unspecified.
+
+    **Remark:** :py:class:`pycsou.opt.solver.stop.StopCriterion_LSQR` is developed for the stopping criterion of LSQR. For
+    computational speed, explicit norms were not computated. Instead, their estimation was used, which is referred from
+    [1]_.
+
+    Examples
+    --------
+    >>> import numpy as np
+     >>> from pycsou.abc import LinOp
+    >>> # Create a PSD linear operator
+    >>> rng = np.random.default_rng(seed=0)
+    >>> mat = rng.normal(size=(10, 10))
+    >>> A = LinOp.from_array(mat).gram()
+    >>> # Create the ground truth 'x_star'
+    >>> x_star = rng.normal(size=(2, 2, 10))
+    >>> # Generate the corresponding data vector 'b'
+    >>> b = A.apply(x_star)
+    >>> # Solve 'Ax=b' for 'x' with the conjugate gradient method
+    >>> lsqr = LSQR(A, show_progress=False)
+    >>> lsqr.fit(b=b)
+    >>> x_solution = lsqr.solution()
+    >>> assert np.allclose(x_star, x_solution)
```

Comment:
Maybe add  some tolerance to not get an error (e.g., `assert np.allclose(x_star, x_solution, rtol=1e-2)`)